ext/ddtrace_profiling_native_extension/heap_recorder.c

```diff
@@ -0,0 +1,489 @@
+int heap_frame_cmp(heap_frame *f1, heap_frame *f2) {
+  int cmp = strcmp(f1->name, f2->name);
+  if (cmp != 0) {
+    return cmp;
+  }
+  cmp = strcmp(f1->filename, f2->filename);
+  if (cmp != 0) {
+    return cmp;
+  }
+  return (int) (f1->line - f2->line);
+}
+
+st_index_t string_hash(char *str, st_index_t seed) {
+  return st_hash(str, strlen(str), seed);
+}
+
+st_index_t heap_frame_hash(heap_frame *frame, st_index_t seed) {
+  st_index_t hash = string_hash(frame->name, seed);
+  hash = string_hash(frame->filename, hash);
+  hash = st_hash(&frame->line, sizeof(frame->line), hash);
+  return hash;
+}
+
+// Important: This should match string_hash behaviour
+st_index_t char_slice_hash(ddog_CharSlice char_slice, st_index_t seed) {
+  return st_hash(char_slice.ptr, char_slice.len, seed);
+}
+
+// ==============
+// Heap Stack API
+// ==============
+heap_stack* heap_stack_init(ddog_prof_Slice_Location locations) {
+  heap_stack *stack = ruby_xcalloc(1, sizeof(heap_stack));
+  stack->frames = ruby_xcalloc(locations.len, sizeof(heap_frame));
+  stack->frames_len = locations.len;
+  for (uint64_t i = 0; i < locations.len; i++) {
+    const ddog_prof_Location *location = &locations.ptr[i];
+    heap_frame *frame = &stack->frames[i];
+    frame->name = ruby_strdup(location->function.name.ptr);
+    frame->filename = ruby_strdup(location->function.filename.ptr);
+    frame->line = location->line;
+  }
+  return stack;
+}
+
+void heap_stack_free(heap_stack *stack) {
+  for (uint64_t i = 0; i < stack->frames_len; i++) {
+    heap_frame *frame = &stack->frames[i];
+    ruby_xfree(frame->name);
+    ruby_xfree(frame->filename);
+  }
+  ruby_xfree(stack->frames);
+  ruby_xfree(stack);
+}
+
+void heap_stack_debug(const heap_stack *stack) {
+  fprintf(stderr, "stack {\n");
+  for (uint64_t i = 0; i < stack->frames_len; i++) {
+    heap_frame *frame = &stack->frames[i];
+    fprintf(stderr, "  frame {\n");
+    fprintf(stderr, "    name: '%s'\n", frame->name);
+    fprintf(stderr, "    filename: '%s'\n", frame->filename);
+    fprintf(stderr, "    line: %lli\n", frame->line);
+    fprintf(stderr, "  }\n");
+  }
+  fprintf(stderr, "}\n");
+}
+
+int heap_stack_cmp(heap_stack *st1, heap_stack *st2) {
+  if (st1->frames_len != st2->frames_len) {
+    return (int) (st1->frames_len - st2->frames_len);
+  }
+  for (uint64_t i = 0; i < st1->frames_len; i++) {
+    int cmp = heap_frame_cmp(&st1->frames[i], &st2->frames[i]);
+    if (cmp != 0) {
+      return cmp;
+    }
+  }
+  return 0;
+}
+
+int heap_stack_cmp_st(st_data_t key1, st_data_t key2) {
+  return heap_stack_cmp((heap_stack *) key1, (heap_stack *) key2);
+}
+
+st_index_t heap_stack_hash(heap_stack *stack, st_index_t seed) {
+  st_index_t hash = seed;
+  for (uint64_t i = 0; i < stack->frames_len; i++) {
+    hash = heap_frame_hash(&stack->frames[i], hash);
+  }
+  return hash;
+}
+
+st_index_t heap_stack_hash_st(st_data_t key) {
+  return heap_stack_hash((heap_stack *) key, FNV1_32A_INIT);
+}
```

Comment:
Minor: I wonder if we should keep the hash somewhere, to avoid recomputing over and over (as it may get a bit expensive)

Reply:
The ruby hash table implementation already keeps this around to avoid recomputations on existing items: https://github.com/ruby/ruby/blob/903b0931a116691386e71fa30fb1698bbd785853/st.c#L133-L137

So this is only called on an external key (which is used on external to st.c lookup/insertion/update/deletion) and these keys would almost always be new objects and thus not have hashes precomputed anyway so I think this is fine?

Reply:
Seems reasonable (and neat that this is done for us!)